dask/array/core.py

```diff
@@ -2642,13 +2707,78 @@ def insert_to_ooc(arr, out, lock=True, region=None):
     dsk = dict()
     for t, slc in zip(core.flatten(arr.__dask_keys__()), slices):
         store_key = (name,) + t[1:]
-        dsk[store_key] = (
-            store_chunk, t, out, slc, lock, region
-        )
+        dsk[store_key] = (store_chunk, t, out, slc, lock, region, keep)
 
     return dsk
 
 
+def load_chunk(x, index, lock, region):
+    """
+    A function inserted in a Dask graph for loading a chunk.
+
+    Parameters
+    ----------
+    x: array-like
+        An array (potentially a NumPy one)
+    index: slice-like
+        Where to store result from ``x`` in ``out``.
+    lock: Lock-like or False
+        Lock to use before writing to ``out``.
+    region: slice-like or None
+        Where relative to ``out`` to store ``x``.
+
+    Examples
+    --------
+
+    >>> a = np.ones((5, 6))
+    >>> load_chunk(a, (slice(None), slice(None)), False, None)  # doctest: +SKIP
+    """
+
+    result = None
+
+    subindex = index
+    if region is not None:
+        subindex = fuse_slice(region, index)
+
+    if lock:
+        lock.acquire()
+    try:
+        result = x[subindex]
+    finally:
+        if lock:
+            lock.release()
+
+    return result
+
+
+def retrieve_from_ooc(keys, dsk):
+    """
+    Creates a Dask graph for loading stored ``keys`` from ``dsk``.
+
+    Parameters
+    ----------
+    keys: Sequence
+        A sequence containing Dask graph keys to load
+    dsk: Mapping
+        A Dask graph corresponding to a Dask Array
+
+    Examples
+    --------
+    >>> import dask.array as da
+    >>> d = da.ones((5, 6), chunks=(2, 3))
+    >>> a = np.empty(d.shape)
+    >>> g = insert_to_ooc(d, a)
+    >>> retrieve_from_ooc(g.keys(), g)  # doctest: +SKIP
+    """
+
+    load_dsk = dict()
+    for each_key in keys:
+        load_key = ('load-%s' % each_key[0],) + each_key[1:]
+        load_dsk[load_key] = (load_chunk, each_key,) + dsk[each_key][3:-1]
```

Comment:
Slight preference to write as a comprehension:

```python
return {('load-%s' % k[0],) + k[1:]: (load_chunk, k,) + dsk[k][3:-1]
        for k in keys}
```

Reply:
Also, can you explain the `[3:-1]` select here in a comment?

Reply:
IDK for some reason that feels less readable to me. However its just personal preference and if there are strong feelings can change.

Yeah this is a little weird. Basically we know we want to keep those argument from the call to `store` previously. Happy to add a comment for it.

Reply:
I personally prefer comprehensions (as long as they're not too complicated), as they make it immediately clear that the loop is just building a dictionary.

A comment would be nice for future maintainers. Thanks.

Reply:
Have left the `for`-loop as is for now.

Added a comment to explain what is happening here. Please let me know your thoughts on it.

Reply:
Just for extra clarity, `dsk[k][3:-1]` translates to `(index, lock, region)` for that `k`.